File: c++/insertSort.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;


void insertSort(vector<int> &arr,int n){
    vector<int> res;
	for(int i = 0; i < n; i++){
        if(res.size() == 0){
            res.push_back(arr[i]);
        }else{
            for(int j = res.size() - 1; j >= 0; j--){
                if(arr[i] >= res[j]){
                    res.insert(res.begin()+j+1,arr[i]);
                    break;
                }else if(j == 0){
                    res.insert(res.begin(),arr[i]);
                }
            }
        }
    }
    for(int i = 0; i < n; i++){
        arr[i] = res[i];
    }
    res.clear();
}

int main(){
	vector<int> nums;
	int len;
	cout<<"input length: "<<endl;
	cin>>len;
	cout<<"input unordered array: "<<endl;
	for(int i = 0; i < len; i++){
		int tmp;
		cin>>tmp;
		nums.push_back(tmp);
	}

	cout<<"before insert sort: "<<endl;
    for(int j = 0; j < len; j++){
    	cout<<nums[j]<<" ";
    }
    cout<<endl;

    insertSort(nums,len);

    cout<<"after insert sort: "<<endl;
    for(int j = 0; j < len; j++){
    	cout<<nums[j]<<" ";
    }
    cout<<endl;
    return 0;
}

File: javascript/insertSort.js
function insertSort(arr){
    let tmp = [];
    for(let i = 0; i < arr.length; i++){
        if(tmp.length == 0) tmp.push(arr[i]);
        else{
            for(let j = tmp.length - 1; j >= 0; j--){
                if(arr[i] >= tmp[j]){
                    insert(tmp,j+1,arr[i]);
                    break;
                }else if(j == 0){
                    insert(tmp,0,arr[i]);
                }
            }
        }
    }
    for(let i = 0; i < arr.length; i++){
        arr[i] = tmp[i];
    }
}

function insert(arr,idx,val){
    if(idx == arr.length){
        arr.push(val);
    }else{
        arr.push(0);
        let tmp = arr[idx];
        for(let j = arr.length - 1; j > idx; j--){
            arr[j] = arr[j - 1];
        }
        arr[idx] = val;
    }

}

let arr = [4,5,1,2,7,5,8,7,9];
console.log("before insert sort",arr);
insertSort(arr);
console.log("after insert sort",arr);